lb: add hash_key support to envoy.lb endpoint metadata

inf-rno opened the pull request:
- Fixes https://github.com/envoyproxy/envoy/issues/13862
- Adds hash_key filter in the envoy.lb metadata key for an endpoint
- If this key is set, then consistent hashing loadbalancers (maglev,
  ring hash) will use its value to hash the corresponding host

Risk Level: Low
Testing: Unit testing
Docs Changes: Unknown
Release Notes: Unknown

inf-rno commented:
Unsure how to document this feature cc @mattklein123

zuercher reviewed (commented):
Thanks. This generally looks good. I'm not sure whether the CI failures are intermittent or something that needs to be fixed on master. If your branch is behind, it might be worth merging master to see if that resolves the problem.

I think we should consider adding an explicit setting for this feature in `config.cluster.v3.Cluster.CommonLbConfig.ConsistentHashingLbConfig`. That will allow operators to explicitly control whether this override happens or not.

As far as documentation goes, `docs/root/intro/arch_overview/upstream/load_balancing/load_balancers.rst` seems appropriate.

inf-rno commented:
I fixed the comments and rebased. I don't think we need another flag in `ConsistentHashingLbConfig`. At that point we might as well stick the whole feature in that struct. This seems more flexible and a good use of the metadata feature.

zuercher commented:
This looks good. Let's get the docs/release notes added.

mattklein123 commented:
Can you merge main to hopefully fix CI and I can take a look?

/wait

mattklein123 commented:
Check CI?

/wait

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/load_balancers.rst:
```diff
@@ -71,7 +71,16 @@ mapped onto a circle (the "ring") by hashing its address; each request is then r
 hashing, and like all hash-based load balancers, it is only effective when protocol routing is used
-that specifies a value to hash on.
+that specifies a value to hash on. If you want something other than the host's address to be used
+as the hash key (e.g. the semantic name of your host in a Kubernetes StatefulSet), then you can specify it 
+in the ``"envoy.lb"``:ref:`filter_metadata <envoy_api_msg_core.Metadata>` e.g.: 
```
Can you actually ref link to the field in the host where this should be set? I think that would be more useful from a docs perspective. Same below and in the release notes.

mattklein123 commented on source/common/upstream/maglev_lb.cc:
```diff
@@ -19,11 +19,10 @@ MaglevTable::MaglevTable(const NormalizedHostWeightVector& normalized_host_weigh
-    ASSERT(!address.empty());
-    table_build_entries.emplace_back(host, HashUtil::xxHash64(address) % table_size_,
-                                     (HashUtil::xxHash64(address, 1) % (table_size_ - 1)) + 1,
+    std::string key_to_hash = hashKey(host, use_hostname_for_hashing);
```
nit: const, same below and in ring code

mattklein123 commented on source/common/upstream/thread_aware_lb_impl.h:
```diff
@@ -26,6 +28,17 @@ class ThreadAwareLoadBalancerBase : public LoadBalancerBase, public ThreadAwareL
+      std::string hash_key = Config::Metadata::metadataValue(
+                                 host->metadata().get(), Config::MetadataFilters::get().ENVOY_LB,
+                                 Config::MetadataEnvoyLbKeys::get().HASH_KEY)
+                                 .string_value();
```
What does this return if the value is not a string? Empty string? Should this be configuration checked somewhere? Or at least have a TODO for that? Please have a test for a non-string key.

inf-rno replied:
id imagine the yaml parser into protobuf would leave the string value empty in that case and it will get handled by the default case (fallback to hostname/resolved address)

mattklein123 replied:
I think that's OK, but do you have a test for this case?

/wait

inf-rno replied:
there are tests for when the metadata is not set and the string is empty e.g. in [maglev ](https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/maglev_lb_test.cc#L101-L169).

i can add another one in HashingLoadbalancer by creating a host with an int_value in the metadata instead of [string_value](https://github.com/envoyproxy/envoy/pull/14162/files#diff-1cf1654ea1233e6793e9a924bbb6ec36beff732bb7ea2c0230d9cc74b87e55f1R128) if you like. But it seems a bit redundant.

mattklein123 replied:
I'm looking for a test in which metadata is set to something that is not a string, as you point out. I don't think this is redundant and it makes sure we don't see any strange behavior with what might be a common config mistake. Thank you!

/wait

mattklein123 reviewed (changes requested):
Thanks generally LGTM with a few small comments. Thank you!

/wait

vanhtuan0409 commented:
Hi @inf-rno can we somehow continue this PR. This feature would greatly increase the flexibility when using envoy for sharding

inf-rno commented:
uhmm. Yes sorry! This fell of my radar for too long. It is basically done except that one test from the review. I'll spend sometime on it today.
